services/httpd/handler.go

```diff
@@ -993,8 +1012,366 @@ func fixLiterals(node influxql.Node) {
+			h.httpError(w, fmt.Sprintf("buckets - user is required to to create bucket %q", brd.Name), http.StatusForbidden)
+			return
+		}
+
+		if dbi == nil {
+			if err := h.QueryAuthorizer.AuthorizeCreateDatabase(user); err != nil {
+				h.httpError(w, fmt.Sprintf("buckets - %q is not authorized to create %q: %s", user.ID(), brd.Name, err.Error()), http.StatusForbidden)
+				return
+			}
+		} else {
+			if err := h.QueryAuthorizer.AuthorizeCreateRetentionPolicy(user, db); err != nil {
+				h.httpError(w, fmt.Sprintf("buckets - %q is not authorized to create %q: %s", user.ID(), brd.Name, err.Error()), http.StatusForbidden)
+				return
+			}
+		}
+	}
+
+	if dbi == nil {
+		if _, err = h.MetaClient.CreateDatabaseWithRetentionPolicy(db, &spec); err != nil {
+			h.httpError(w, fmt.Sprintf("buckets - cannot create bucket %q: %s", brd.Name, err.Error()), http.StatusBadRequest)
+			return
+		}
+	} else {
+		if _, err = h.MetaClient.CreateRetentionPolicy(db, &spec, false); err != nil {
+			h.httpError(w, fmt.Sprintf("buckets - cannot create bucket %q: %s", brd.Name, err.Error()), http.StatusBadRequest)
+			return
+		}
+	}
+}
+
+func (h *Handler) serveBucketDeleteV2(w http.ResponseWriter, r *http.Request, user meta.User) {
+	db := r.URL.Query().Get(":db")
+	rp := r.URL.Query().Get(":rp")
+	id := fmt.Sprintf("%s/%s", db, rp)
+
+	if h.Config.AuthEnabled {
+		if user == nil {
+			h.httpError(w, fmt.Sprintf("delete bucket - user is required to delete from database %q", db), http.StatusForbidden)
+			return
+		}
+
+		if err := h.QueryAuthorizer.AuthorizeDeleteRetentionPolicy(user, db); err != nil {
+			h.httpError(w, fmt.Sprintf("delete bucket - %q is not authorized to delete %q: %s", user.ID(), id, err.Error()), http.StatusForbidden)
+			return
+		}
+	}
+
+	if dbi := h.MetaClient.Database(db); dbi == nil {
+		h.httpError(w, fmt.Sprintf("delete bucket - not found: %q", id), http.StatusNotFound)
+		return
+	} else if rpi := dbi.RetentionPolicy(rp); rpi == nil {
+		h.httpError(w, fmt.Sprintf("delete bucket - not found: %q", id), http.StatusNotFound)
+		return
+	} else if err := h.Store.DeleteRetentionPolicy(db, rp); err != nil {
+		h.httpError(w, fmt.Sprintf("delete bucket %q: %s", id, err.Error()), http.StatusBadRequest)
+		return
+	} else if err := h.MetaClient.DropRetentionPolicy(db, rp); err != nil {
+		h.httpError(w, fmt.Sprintf("delete bucket %q: %s", id, err.Error()), http.StatusBadRequest)
+		return
+	}
+}
+
+func (h *Handler) serveRetrieveBucketV2(w http.ResponseWriter, r *http.Request, user meta.User) {
+	// This is the API for returning a single bucket
+	// In V1, BucketID and BucketName are the same: "db/rp"
+	db := r.URL.Query().Get(":db")
+	rp := r.URL.Query().Get(":rp")
+	if h.Config.AuthEnabled {
+		if user == nil {
+			h.httpError(w, fmt.Sprintf("retrieve bucket - user is required for database %q", db), http.StatusForbidden)
+			return
+		}
+
+		if err := h.QueryAuthorizer.AuthorizeDatabase(user, influxql.ReadPrivilege, db); err != nil {
```

Comment:
Does the error message change depending on whether the bucket exists or not?

Reply:
Not in OSS.  Always returns 
`fmt.Sprintf("user %q, requires %s for database %q", u.ID(), priv.String(), database)`